src/backend/access/index/genam.c

```diff
@@ -606,3 +609,127 @@ systable_endscan_ordered(SysScanDesc sysscan)
+ *
+ * Overwriting violates both MVCC and transactional safety, so the uses of
+ * this function in Postgres are extremely limited.  Nonetheless we find some
+ * places to use it.  See README.tuplock section "Locking to write
+ * inplace-updated tables" and later sections for expectations of readers and
+ * writers of a table that gets inplace updates.  Standard flow:
+ *
+ * ... [any slow preparation not requiring oldtup] ...
+ * systable_inplace_update_begin([...], &oldtup, &inplace_state);
+ * if (!HeapTupleIsValid(oldtup))
+ *	elog(ERROR, [...]);
+ * tup = heap_copytuple(oldtup);
+ * ... [buffer is exclusive-locked; mutate "tup"] ...
+ * if (dirty)
+ *	systable_inplace_update_finish(inplace_state, oldtup, tup);
+ * else
+ *	systable_inplace_update_cancel(inplace_state, oldtup);
+ *
+ * The first several params duplicate the systable_beginscan() param list.
+ * "p_oldtup" is an output parameter, assigned NULL if the key ceases to
+ * find a live tuple.  (In PROC_IN_VACUUM, that is a low-probability transient
+ * condition.)  If "p_oldtup" gets non-NULL, you must make a copy of it and
+ * pass output parameter "state" to systable_inplace_update_finish() or
+ * systable_inplace_update_cancel().
+ */
+HeapTuple
+systable_inplace_update_begin(Relation relation,
+							  Oid indexId,
+							  bool indexOK,
+							  Snapshot snapshot,
+							  int nkeys, const ScanKeyData *key,
+							  HeapTuple *p_oldtup,
+							  void **state)
+{
+	Assert(nkeys > 0);
+	ScanKey		mutable_key = palloc(sizeof(ScanKeyData) * nkeys);
+	int			retries = 0;
+	SysScanDesc scan;
+	HeapTuple	oldtup;
+
+	/*
+	 * Accept a snapshot argument, for symmetry, but this function advances
+	 * its snapshot as needed to reach the tail of the updated tuple chain.
+	 */
+	Assert(snapshot == NULL);
+
+	Assert((relation->rd_id == RelationRelationId ||
+			relation->rd_id == DatabaseRelationId) || !IsSystemRelation(relation));
+
+	/* Loop for an exclusive-locked buffer of a non-updated tuple. */
+	for (;;)
+	{
+		CHECK_FOR_INTERRUPTS();
+
+		/*
+		 * Processes issuing heap_update (e.g. GRANT) at maximum speed could
+		 * drive us to this error.  A hostile table owner has stronger ways to
+		 * damage their own table, so that's minor.
+		 */
+		if (retries++ > 10000)
+			elog(ERROR, "giving up after too many tries to overwrite row");
+
+		memcpy(mutable_key, key, sizeof(ScanKeyData) * nkeys);
+		scan = systable_beginscan(relation, indexId, indexOK, snapshot,
+								  nkeys, mutable_key);
+		oldtup = systable_getnext(scan);
+		if (!HeapTupleIsValid(oldtup))
+		{
+			systable_endscan(scan);
+			*p_oldtup = NULL;
+			return NULL;
+		}
+
+		Assert(scan->scan || scan->iscan);
+		if (heap_inplace_lock(scan->heap_rel,
+							  oldtup, scan->scan ? scan->scan->rs_cbuf : scan->iscan->xs_cbuf))
+			break;
+		systable_endscan(scan);
+	};
+
+	// The caller must make a copy of the tuple.
+	*p_oldtup = oldtup;
+	*state = scan;
+
+	return heap_copytuple(oldtup);
```

Comment:
Maybe do it the other way around? Write the copy to the argument, as done in the original commit, and just return the original old tuple?